lib/checkother.cpp

```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
         // Skip if address is taken
         if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            //Is it the right side of an initialization of a non-const reference
+            bool usedInAssignment = [&]
+                {
+                    auto findAssignment = [&](const Token* t) { return Token::findmatch(t, "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId()); };
+                    for (const Token* match = findAssignment(var->nameToken()); match != nullptr; match = findAssignment(match->next()))
+                    {
+                        const Variable* matchvar = match ? match->tokAt(2)->variable() : nullptr;
+                        if (matchvar && !matchvar->isConst())
+                            return true;
+                    }
+                    return false;
+                }();
+            if (usedInAssignment)
+                continue;
+        }
+        // Skip if we ever cast this variable to a non-const type
+        {
+            bool castToNonConst = [&]
+                {
+                    for (const Token* tok = var->nameToken(); tok != scope->bodyEnd && tok != nullptr; tok = tok->next())
+                    {
+                        if (tok->isCast())
+                        {
+                            if (Token::simpleMatch(tok->previous(), ">"))
+                            {
+                                // dynamic_cast<X&>(y)
+                                //                 ^    we're at this operator
+                                bool castToConst = Token::simpleMatch(tok->previous()->link()->next(), "const");
```

Comment:
can you please try to use the ValueType:

    if (tok->valueType() && (tok->valueType()->constness & 1))

I am not sure if we want to test `constness & 1` or `constness > 1`. What should happen for `dynamic_cast<X * const>`?

Reply:
I don't think ```dynamic_cast<X * const>``` makes any sense. It's already an rvalue. What difference does it make if it's const? You can freely say ```B* b = dynamic_cast<B * const>(x)``` but can't say ```B* b = dynamic_cast<B const *>(x)```

It's not clear to me what valueType represents or how the bits of constness should be interpreted. Can you explain?